[NEW] Cyclic boosting interface

setoguchi-naoki opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://skpro.readthedocs.io/en/latest/contribute.html
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
#132 


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
J-qpd estimation by quantile regression with Cyclic boosting, A output distribution's parameters are given by j-qpd's loc and scale

#### Does your contribution introduce a new dependency? If yes, which one?
cyclic_boosting package
<!--
If your contribution does add a new core dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core skpro package to a minimum.
-->

#### What should a reviewer concentrate their feedback on?
Please check my probability distribution prediction manner is right or not

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?
Not yet

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).


<!--
Thanks for contributing!
-->

fkiraly commented:
looks good!

Did this get closed by accident?

setoguchi-naoki commented:
@fkiraly 
Hi, Franz, I'm one of the contributor of Cyclic boosting. I closed the PR manually so that I wanted to get review from another contributor of Cyclic boosting on my forked repository because my code is a draft.

fkiraly commented:
> I closed the PR manually so that I wanted to get review from another contributor of Cyclic boosting on my forked repository because my code is a draft.

My opinion, it's best to have reviews here - general best practice recommendation to have all discussion in the "host repo" and not in the fork, as the host repo is indexed by various tools, while fork repos are in general not (and can also disappear more easily).

It's marked as draft too, so no danger of getting it merged.

Also feel free to ask any of the skpro contribs for review!

setoguchi-naoki commented:
@fkiraly 
Thank you for giving me your opnion and follow your comment. I have finished to implement for interface of Cyclic boosting, so Would you review the code for me? I concern weather my implementation fit skpro's manner or not.

FelixWick commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,315 @@
+
+        return self
+
+    def _predict(self, X):
```
I would suggest to not run the full quantile regression - QPD thing here, but simply use a mean prediction mode (e.g., pipeline_CBPoissonRegressor for a multiplicative model on range 0, inf).

FelixWick replied:
But that would require two different fit modes (as we want to use quantile regression with subsequent QPD for predict_proba), from which the user would have to select via a parameter in the init of the estimator. So, I can understand your rationale of using the mean of QPD. If we wanted to keep a single fit mode, we might consider using just the median (anyway predicted by the middle quantile regression model for QPD) though (avoiding the complexity via the mean of QPD).

setoguchi-naoki replied:
@FelixWick 
How about Having two kind of estimators in Cyclic boosting class? I think it look seemless from user because keeping the single fit mode. Is it bit complicated?

FelixWick replied:
Yes, that is what I meant. But at fit time, the user needs to decide what estimator to train. It would be a waste of runtime to train both always. (That is also true for the quantile regression of three different quantiles for the QPD, if all you want is the median.)

fkiraly replied:
is it perhaps helpful to look at the QR model by @Ram0nB here?
https://github.com/sktime/skpro/blob/main/skpro/regression/multiquantile.py

FelixWick replied:
Ok, I changed my mind. For simplicity, what do you think of the following structure:
- always fit three quantiles: alpha, 0.5 (median), and 1-alpha (alpha to be defined as parameter instead of quantiles triplet). That means just use quantile regression as single estimator. But we should use either multiplicative or additive mode, depending on the target range.
- in predict_proba, give QPD as distribution.
- in predict, give the median as point estimate.
- in predict_quantiles, give the three quantile estimates (alpha, median, 1-alpha), or, if that's not possible from the interface, just the alpha estimate.

setoguchi-naoki replied:
@FelixWick 
I followed your comments. Additionally, predict_interval method was implemented. In predict_quantiles, if alphas are same with the quantiles that the estimator train, the method returns estimated value directlly,  if it is not same, the method returns specified quantile value from QPD.

FelixWick commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,315 @@
+            self.quantile_values.append(yhat)
+
+        j_qpd = J_QPD_S(
+            0.2,
```
I would suggest to make this a parameter of the estimator, 0.2 is an arbitrary choice.

FelixWick commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,245 @@
+from skpro.distributions.base import BaseDistribution
+
+
+class J_QPD_S(BaseDistribution):
```
We might also add the bounded version J_QPD_B.

FelixWick commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,315 @@
+            self.quantile_values.append(yhat)
+
+        j_qpd = J_QPD_S(
+            0.2,
```
This should be self.quantiles[0], as it needs to fit the lower quantile in the estimator (which is a parameter above).

setoguchi-naoki replied:
@FelixWick 
Does alpha always mention that it should be lower quantile, right? the alpha is fixed as 0.2 (20 percentile) in the current code
`self.quantiles = [0.2, 0.5, 0.8]`
and
`j_qpd = QPD_S(
            alpha=0.2,
            qv_low=self.quantile_values[0],
            qv_median=self.quantile_values[1],
            qv_high=self.quantile_values[2],
            index=index,
            columns=y_cols,
        )`

Do you mean the change like this? (and add alpha as a class's parameter)
`self.quantiles = [alpha, 0.5, 1 - alpha]`
and
`j_qpd = QPD_S(
            alpha=alpha,
            qv_low=self.quantile_values[0],
            qv_median=self.quantile_values[1],
            qv_high=self.quantile_values[2],
            index=index,
            columns=y_cols,
        )`

FelixWick replied:
No, no need to introduce a new parameter, you can simply use self.quantiles[0], this corresponds to the alpha in QPD class.

FelixWick commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,245 @@
+
+    def pdf(self, x):
+        # cdf -> pdf calculation because j-qpd's pdf calculation is bit complex
+        def cdf_func(x, theta, kappa, delta, n):
```
Call the function below to avoid code duplication here.

FelixWick commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,245 @@
+    def pdf(self, x):
+        # cdf -> pdf calculation because j-qpd's pdf calculation is bit complex
+        def cdf_func(x, theta, kappa, delta, n):
+            cdf_arr = self.phi.cdf(
```
Try to avoid the code duplication here.

FelixWick replied:
And in other places :).

setoguchi-naoki replied:
I have avoided some duplication and relocated

FelixWick commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,315 @@
+
+        if self.distr_type == "QPD":
+            return self.qpd
+        else:
```
I don't think we should offer this option, matching the QPD mean and variance on Gauss or Laplace distributions is odd at best (and can be severely off).

fkiraly replied:
up to you

setoguchi-naoki replied:
I removed these options

FelixWick commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,315 @@
+        y_pred = pd.DataFrame(j_qpd.mean(), index=index, columns=y_cols)
+        return y_pred
+
+    def _predict_proba(self, X):
```
I would suggest to also use predict_quantiles for the quantile predictions. And here we can then just give the QPD.

fkiraly replied:
hm, that's an interesting question. `predict_quantiles` will automatically default to the `quantiles` method of the `predict_proba` return, that ensures consistency. The interface does not enforce exact consistency, so if there's two consistent options, a method/strategy parameter in the `__init__` may be the way to go.

FelixWick replied:
Ok, I see. In that case, let's stay consistent. Let me summarize my suggestion in the other comment thread :).

FelixWick commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,245 @@
+        self.delta = 1.0 / self.c * sinh(arccosh((self.H - self.L) / (2 * _min)))
+        self.kappa = 1.0 / (self.delta * self.c) * _min
+
+        # DataFrame
```
I don't get this index thing here.

fkiraly replied:
yes, this does not look right to me. The index etc usually only "lives" in the distribution itself, not in the parameters.

FelixWick commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,245 @@
+            return cdf_arr
+
+        pdf = []
+        for i in x.index:
```
Aren't these values all the same?

setoguchi-naoki replied:
X is pandas Datafreme which include a point of distribution as value and corresponded QPD's index in my code. So,  these have different value and pdf method returns specific pdf value. However, I'm not sure that my implementation is correct. Please point out my understanding or implementation's mistake if it is exist.

fkiraly commented:
I'm noticing the build failure in readthedocs, `exception: no module named skpro.regression.cyclic_boosting`

This usually indicates some mixup with imports, although I was unable to find any error in a couple minutes searching.

fkiraly commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,245 @@
+"""J-QPD probability distribution."""
```
hm, only public functions, though.

setoguchi-naoki replied:
removed copying code and import J_QPD_S, J_QPD_B from cyclic_boosting package

fkiraly commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,315 @@
+"""
+
+"""License.
+EPL 2.0
```
The interfaces in sktime/skpro should be released under the BSD-3 license of the repo.
Of course the `cyclic_boosting` package can be subject to a different license.

I would think this should be ok, as there is no substantial logic, only interface logic, in the `skpro` module? Comments appreciated, especially if you think otherwise.

Btw, the docstring of the class is what this appears as in the doc, so I would recommend to give more credit to the upstream package, see for instance the `mapie` regressor.

FelixWick replied:
I agree. As I also commented before, we should not copy code from Cyclic Boosting here, but just call the respective functionality.

setoguchi-naoki commented:
@fkiraly 
> I'm noticing the build failure in readthedocs, exception: no module named skpro.regression.cyclic_boosting
This usually indicates some mixup with imports, although I was unable to find any error in a couple minutes searching.

I got this error message, I don't know why it is occured. Do you have any ideas?
Cyclic boosting's version is correct.
> ModuleNotFoundError: CyclicBoosting requires package 'cyclic_boosting>=1.2.1' to be present in the python environment, with version >=1.2.1, but incompatible version 1.0 was found

fkiraly commented:
hmmmm, @setoguchi-naoki, the typical source of the error message would be a combination fo two things:

* the lower bound `"cyclic-boosting>=1.2.1"` that you put in the `python_dependencies` tag
* `cyclic-boosting 1.0` being actually present in the python env.

Which is strange, because the `pyproject.toml` contains `cyclic-boosting>=1.2.1` as well.

Are you getting this on local tests? If so, your environment might be at fault.

fkiraly commented:
PS: the code-quality check is coming up with a few linting issues, let us know if you want advice on setting up local linting!

This usually runs to prevent syntax errors going through, but code formatting issues will also prevent remote CI from running.

setoguchi-naoki commented:
@fkiraly 
Sorry for being late. I got the Import error's reason, there is a little mistake in Cyclic boosting package. the problem will be solved soon. However, I got the other import error in part of docstring's test

> sphinx.ext.autosummary.ImportExceptionGroup: no module named skpro.distributions.Empirical

Could you please support for me?

setoguchi-naoki commented:
@FelixWick 
Thank you for your review, I implemented extended J_QPDs. The problem which I have is only the issue related to build docstring but I don't know how to solve it even though reading its error code. :(

FelixWick commented:
@setoguchi-naoki Maybe related to this? https://stackoverflow.com/questions/62226506/python-sphinx-autosummary-failed-to-import-module

setoguchi-naoki commented:
I tried to build docstring on my local machine. it run with some warnings but it looks successful to me
<img width="292" alt="screen_shot 2023-12-26 141536" src="https://github.com/sktime/skpro/assets/104494531/d50d2d8c-3e28-42c4-80ae-e24cc36866ae">

fkiraly commented:
Apologies for the delay, I was busy at work and with the `sktime` releases.

The docs failure is actually a genuine test failure, due to how `scikit-base` based packages like `skpro`, `sktime` handle soft dependencies.

In order to not require of every user to install all the soft dependencies, they must be isolated, that is, the package should install and not break with it is not present.

For this, imports of soft dependencies, like `cyclic_boosting`, need to reside inside classes or methods, not at the top of the file.

General instructions here:
https://www.sktime.net/en/latest/developer_guide/dependencies.html#handling-soft-dependencies

But I'll also point out what to do in the PR.

fkiraly commented on skpro/distributions/qpd.py:
```diff
@@ -0,0 +1,705 @@
+
+import numpy as np
+import pandas as pd
+from cyclic_boosting.quantile_matching import (
```
these imports should move into the distribution classes.

My recommendation is to move them inside the methods where they are needed.

fkiraly replied:
done in my commits

fkiraly commented on skpro/regression/cyclic_boosting.py:
```diff
@@ -0,0 +1,428 @@
+
+import numpy as np
+import pandas as pd
+from cyclic_boosting import (
```
similar, these imports should move into the class methods

fkiraly replied:
done in my commits

fkiraly reviewed (changes requested):
Nice!

Apologies for the hiatus, that was work related (and I had to finish the `sktime` releases).

Looks good to go, I see only an issue with dependency isolation, as mentioned above. That should also solve the docs issue.

fkiraly commented:
@setoguchi-naoki, @FelixWick, I proceeded pushing some commits which isolate the dependencies - I hope this is ok, if not, please revert.

What I did:

* move imports into classes and methods
* skip the class doctest, move to a pytest test that is only executed when the required dependencies are present

fkiraly reviewed (changes requested):
The failures now are genuine.

The reason seems to be, `CyclicBoosting` does not construct with all the parameter sets returned by `get_test_params`.

In addition, the docstring examples of the distributions should be isolated, similar to the `CyclicBoosting` one (I skipped it and moved to conditional pytest, in `test_cyclic_boosting`)

(I will stop here with contributions, as most remaining issues touch logic which you are more familiar with)

fkiraly commented:
@setoguchi-naoki, @FelixWick, let us know if you need a hand!

setoguchi-naoki commented:
@fkiraly 
Thank you for your cooperation! I solved some bug and add a new test script for qpd.py. then, all check have passed. please check them.

fkiraly reviewed (changes requested):
Nice! Looks almost done, failures are two kinds:

* the `cyclic_boosting` package appears not to be python 3.12 compatible, so the `python_version` tag should be set to `<3.12`
* at one point, the inner logic seems to be looking for a column `s1` in `X` that does not exist - seems like a column name treatment bug.

fkiraly reviewed (approved):
nice - looks all good now!

You should also add yourself to the `.all-cotributorsrc` file!

fkiraly commented:
(tests have not run through yet, but I've started them again)

setoguchi-naoki commented:
@fkiraly 
I have found why that error occued. Test dataset's columns are updated at line 156 in scenarios_regressor_proba.py.
`X.columns = ["age", "sex", "bmi", "bp", 1, 2, 3, 4, 5, 6]` 
It needs to same feature_propety's key names and dataset columns for Cyclic boositing but the former is fixed as a test paramter now. How can I change dynamically test parameters along with changing scenario?

fkiraly reviewed (changes requested):
Oh, I see!
Let me rephrase if I understood correctly:

`CyclicBoosting` has a parameter, `feature_properties`, in which keys need to be the same as (all) variable names. That causes a failure in the test, because the same parameter set is tested with all data scenarios, where there are different sets of variable names.

This is an interesting question - optimally, the user does not have to worry about variable specific parameters.

May I suggest to auto-detect the types? A simple solution could be, use `pandas` [`is_float_dtype`](https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_float_dtype.html) on `X.dtype`. For each column where this is true, assume continuous, otherewise categorical. I would suggest to do that if `feature_properties` is `None`. This would work as follows - you have an internal `_feature_properties` (with underscore at the start), and if `feature_properties` is given, it is that, and if not, it is the auto-detected one.

That would solve the issue with the tests, and imo would make the algorithm much more convenient for the user too.

Further requests relate to the docstring of `CyclicBoosting`:

* it should give credit to the package, e.g., say sth like "direct interface to `cyclic_boosting classes etcetc`". I would do that in the second line.
* `feature_properties` should be listed in the docstring, or it might be confusing to a user. Or is the specification very long?
* All parameters with a default should have that default stated in the docstring.

FelixWick commented:
In fact, the user doesn't need to give feature properties, there is a default mechanism (which, admittedly, still leaves some room for improvement ;)). So, your suggestion should work out-of-the-box, I think.

fkiraly commented:
Nice!
Still one would have to set a default, right now the user is forced to specify.

I would then advice to remove the variable specific parameters from `get_test_params`, and add separate `pyteset` based tests to ensure correctness when they are passed.

FelixWick commented:
Let me know if you need my help, @setoguchi-naoki

setoguchi-naoki commented:
@FelixWick @fkiraly 
Thank you guys for your suggestion! I will make simple feature analysis function and add test for it. In fact, I am already developing automatic feature type analysis function for Cyclic boosting package. It can detect but simple data property also more complicate ones (e.g. seasonality). However, I guess it need to have a little time to release, please allow me to take simple approach for now. I will reflect it for skpro after releasing this function.

setoguchi-naoki commented:
@fkiraly 
As Felix said, feature_properties (and feature_groups) always don't need to set up. So I just updated CyclicBoosting class's paramaters and removed specific test paramaters as a most simple solution. if feature_properties and feature_groups are None, these are set automatically in cyclic boosting model. I guess it should passed all test by this modification.

fkiraly reviewed (changes requested):
Excellent, thanks!

Could you perhaps add a test that uses the manual feature typing to `test_cyclic_boosting`, so we have that covered?

Otherwise, I have no additional requests, very nice.

Once you think you are ready, pleas switch this PR to "ready for review".

setoguchi-naoki commented:
I have finished :)

fkiraly reviewed (approved):
Nice!
All green for merge if tests pass.

fkiraly reviewed (changes requested):
Fails with `test_cyclic_boosting_with_manual_paramaters - ValueError: The SPT values need to be monotonically increasing.` - error seems genuine.

FelixWick commented:
The quantile predictions need to be ordered correctly for the QPD to make sense. And as we are dealing with statistical models, it can happen that this is not the case, especially in low-statistics tests. So, I would suggest to just tune the respective test ;).

fkiraly commented:
that, or adding a coercion?

FelixWick commented:
Yeah, I have implemented a potential coercion in QPD_RegressorChain in Cyclic Boosting. But a simpler approach would do as well here, I think.

setoguchi-naoki commented:
I modified SPT's value checking from replacement by mean to sorting.

fkiraly commented:
Seems lik there are sporadic timeouts of `test_cyclic_boosting_with_manual_paramaters` - a single test run is limited to 10min or it will fail with the assumption that the long runtime is a consequence of a failure.

setoguchi-naoki commented:
All test have passed :)

fkiraly reviewed (approved):
🎉 🎆 🎉 🎆 🎉

fkiraly commented:
Minor question, non-blockinng, should the QPD not go in the "non-parametric" section of the distirbutions?

They are called "quatile parameterized" but afaik are typically considered non-parametric since they are explicit in terms of quantiles (which are interpolated) rather than intrinsic simple parameters.

fkiraly commented:
Just looking through open convos at the moment, and I found this by @FelixWick:

> But as I wrote below, I think we should rename the parameter alpha in predict_quantiles, because it is confusing.

Apologies, I overlooked this - are you suggesting that the interface be changed in general for all estimators, with a rename of `alpha`? There might be a case for that, though I cannot find the arguments you are referencing.

FelixWick commented:
> Minor question, non-blocking, should the QPD not go in the "non-parametric" section of the distributions?
> 
> 
> 
> They are called "quantile parameterized" but afaik are typically considered non-parametric since they are explicit in terms of quantiles (which are interpolated) rather than intrinsic simple parameters.

Yes, that sounds reasonable.

FelixWick commented:
> Just looking through open convos at the moment, and I found this by @FelixWick:
> 
> 
> 
> > But as I wrote below, I think we should rename the parameter alpha in predict_quantiles, because it is confusing.
> 
> 
> 
> Apologies, I overlooked this - are you suggesting that the interface be changed in general for all estimators, with a rename of `alpha`? There might be a case for that, though I cannot find the arguments you are referencing.

No, that was not a general comment, just for Naoki's implementation. But I think he fixed it. Right, @setoguchi-naoki ?

fkiraly commented:
>> Minor question, non-blocking, should the QPD not go in the "non-parametric" section of the distributions?

> Yes, that sounds reasonable.

Ok, moving to the non-parametric taxon then.

setoguchi-naoki commented:
> > Just looking through open convos at the moment, and I found this by @FelixWick:
> > > But as I wrote below, I think we should rename the parameter alpha in predict_quantiles, because it is confusing.
> > 
> > 
> > Apologies, I overlooked this - are you suggesting that the interface be changed in general for all estimators, with a rename of `alpha`? There might be a case for that, though I cannot find the arguments you are referencing.
> 
> No, that was not a general comment, just for Naoki's implementation. But I think he fixed it. Right, @setoguchi-naoki ?

Yes, it is only for my predict_quantiles implementation now.
